qdrant-landing/content/articles/io_uring.md

```diff
@@ -0,0 +1,196 @@
+one system call per operation, which incurs a context switch - an operation
+which is in itself not that slow (basically registers are stored and loaded
+from a known good location), but the memory accesses required led to longer
+and longer wait times for the CPU.
+
+### Memory-mapped IO
+
+Another way of dealing at least with file IO (which unlike network IO doesn't
+usually have a hard time requirement) is to map parts of files into memory -
+basically the system fakes having that chunk of the file in memory, so when
+you read from a location there, the kernel interrupts you to load the needed
+data from disk, and then sends you on your merry way, whereas writing to the
+memory will also notify the kernel. Also the kernel can prefetch stuff while
+the program is running, thus reducing the likelyhood of interrupts.
+
+Thus there is still some overhead, but (especially in asynchronous
+applications) it's far less than with `epoll`. The reason this API is rarely
+used in web servers is that these usually have a large variety of files to
+access (unlike a database, which can map its own backing store into memory
+once).
+
+### Combating the Poll-ution
+
+There were multiple experiments (open source or proprietary) to improve
+matters, some even going so far as moving a HTTP server into the kernel
+(which of course brought its own share of problems). Others like intel added
+their own APIs that ignored the kernel and worked directly on the hardware
+(on intel hardware, that is).
+
+Finally Jens Axboe took matters into his own hands and proposed a ring buffer
+based interface called *io\_uring*. The buffers are not directly for data, but
+for operations. User processes can setup a Submission Queue (SQ) and a
+Completion Queue (CQ), both of which are shared between the process and the
+kernel, so there's no copying overhead.
+
+![io_uring diagram](/articles_data/io_uring/io_uring_diagram.svg)
+
+Apart from avoiding copying overhead, the queue-based architecture lends
+itself to multithreading (as item insertion/extraction can be made lockless),
+and once the queues are set up, there is no further syscall that would stop
+any user thread.
+
+Servers that use this can easily get to over 100k concurrent requests. Today
+Linux allows asynchronous IO via io\_uring for network, disk and accessing
+other ports (e.g. for printing or recording video).
+
+## And what about Qdrant?
+
+Qdrant can store everything in memory, but not all data sets may fit, which can
+require storing  on disk. Before io\_uring, Qdrant used mmap to do its IO. This
+leads to some modest overhead in case of disk latency, because the kernel may
+stop a user thread trying to access a mapped region, which incurs some context
+switching overhead plus the wait time until the disk IO is finished. All in all,
+this works reasonably well with the asynchronous nature of Qdrant's core.
+
+One of the great optimization tricks Qdrant pulls is quantization (either
```

Comment:
We should really try and use professional wording. Qdrant doesn't pull tricks. It offers functionalities so that users can search and manipulate their data in better ways.